kratos/includes/properties.h

```diff
@@ -278,31 +293,383 @@ class Properties : public IndexedObject
+            return true;
+        }
+
+        // Looking into the database
+        auto property_iterator = mSubPropertiesList.find(SubPropertyIndex);
+        if (property_iterator != mSubPropertiesList.end()) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * @brief This method checks if the subproperty exists from the index corresponding to the property id  (constant version)
+     * @param SubPropertyIndex The index of the subproperty to be get
+     * @return True if there is such subproperty, false otherwise
+     */
+    bool HasSubPropertiesById(const IndexType SubPropertyIndex) const
+    {
+        // Checking if the id is the current properties
+        if (this->Id() == SubPropertyIndex) {
+            KRATOS_WARNING("Properties") << "There is no subproperty with the Id: " << SubPropertyIndex << ". The current property has that Id" << std::endl;
+            return true;
+        }
+
+        // Looking into the database
+        auto property_iterator = mSubPropertiesList.find(SubPropertyIndex);
+        if (property_iterator != mSubPropertiesList.end()) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * @brief This method gets the subproperty from the index corresponding to the property id
+     * @param SubPropertyIndex The index of the subproperty to be get
+     * @return The pointer to the subproperty of interest
+     */
+    Properties::Pointer pGetSubPropertiesById(const IndexType SubPropertyIndex)
+    {
+        // Checking if the id is the current properties
+        if (this->Id() == SubPropertyIndex) {
+            return Kratos::make_shared<Properties>(*this);
+        }
+
+        // Looking into the database
+        auto property_iterator = mSubPropertiesList.find(SubPropertyIndex);
+        if (property_iterator != mSubPropertiesList.end()) {
+            return *(property_iterator.base());
+        } else {
+            KRATOS_WARNING("Properties") << "Subproperty ID: " << SubPropertyIndex << " is not defined on the current Properties ID: " << this->Id() << " creating a new one with ID: " << SubPropertyIndex << std::endl;
+            auto p_new_property = Kratos::make_shared<Properties>(SubPropertyIndex);
+            AddSubProperty(p_new_property);
+            return p_new_property;
+        }
+    }
+
+    /**
+     * @brief This method gets the subproperty from the index corresponding to the property id (constant version)
+     * @param SubPropertyIndex The index of the subproperty to be get
+     * @return The pointer to the subproperty of interest
+     */
+    Properties::Pointer pGetSubPropertiesById(const IndexType SubPropertyIndex) const
+    {
+        // Checking if the id is the current properties
+        if (this->Id() == SubPropertyIndex) {
+            return Kratos::make_shared<Properties>(*this);
```

Comment:
Why do you do this?

Reply:
Because I have not understood that the Id could be duplicated